src/Common/src/Interop/Windows/winhttp/Interop.SafeWinHttpHandle.cs

```diff
@@ -11,30 +12,51 @@ internal partial class Interop
     internal partial class WinHttp
     {
         // Issue 2501: Fold SafeWinHttpHandleWithCallback and SafeWinHttpHandle into a single type.
-        //
-        // SafeWinHttpHandleWithCallback is a better pattern; SafeWinHttpHandle is probably a better
-        // name. The implementation for the former should probably be folded into the latter.
-
+        // SafeWinHttpHandleWithCallback is incorrectly overriding the Dispose(bool disposing) method
+        // and will be fixed as part of merging the classes.
         internal class SafeWinHttpHandle : SafeHandleZeroOrMinusOneIsInvalid
         {
+            private SafeWinHttpHandle _parentHandle = null;
+            
             public SafeWinHttpHandle() : base(true)
             {
             }
 
-            public static void DisposeAndClearHandle(ref SafeWinHttpHandle winHttpHandler)
+            public static void DisposeAndClearHandle(ref SafeWinHttpHandle safeHandle)
             {
-                if (winHttpHandler != null)
+                if (safeHandle != null)
                 {
-                    winHttpHandler.Dispose();
-                    winHttpHandler = null;
+                    safeHandle.Dispose();
+                    safeHandle = null;
                 }
             }
 
+            public void SetParentHandle(SafeWinHttpHandle parentHandle)
+            {
+                Debug.Assert(_parentHandle == null);
+                Debug.Assert(parentHandle != null);
+                Debug.Assert(!parentHandle.IsInvalid);
+
+                bool success = false;
+                parentHandle.DangerousAddRef(ref success);
+                if (!success)
```

Comment:
@stephentoub Yes, the docs are confusing; your interpretation is correct.

This weird pattern was invented for thread abort hardening (on desktop). Thread abort exception can be injected anywhere. The "ref bool success" was intended to make it easy for you cleanup code to determine whether to call DangerousRelease.

The [documentation](https://msdn.microsoft.com/en-us/library/dd289498(v=vs.110).aspx) for Monitor.Enter(object, ref bool) - that uses the same pattern - has more detailed description.
 
- The check for !success in this code can be deleted - it won't be ever executed.
- You do not need to do anything else since we are not doing thread abort hardening in .NET Core. If you would like to write it hardened, you would need to add bool field to your object; and cleanup in finalizer only if it is set.